zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java

```diff
@@ -0,0 +1,295 @@
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.backup;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Range;
+import org.apache.zookeeper.server.backup.BackupUtil.BackupFileType;
+import org.apache.zookeeper.server.backup.BackupUtil.ZxidPart;
+import org.apache.zookeeper.server.backup.exception.RestoreException;
+import org.apache.zookeeper.server.backup.storage.BackupStorageProvider;
+import org.apache.zookeeper.server.backup.storage.BackupStorageUtil;
+import org.apache.zookeeper.server.persistence.FileTxnSnapLog;
+import org.apache.zookeeper.server.persistence.Util;
+import org.apache.zookeeper.server.util.ZxidUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *  This is a tool to restore, from backup storage, a snapshot and set of transaction log files
+ *  that combine to represent a transactionally consistent image of a ZooKeeper ensemble at
+ *  some zxid.
+ */
+public class RestoreFromBackupTool {
+  private static final Logger LOG = LoggerFactory.getLogger(RestoreFromBackupTool.class);
+
+  BackupStorageProvider storage;
+  FileTxnSnapLog snapLog;
+  long zxidToRestore;
+  boolean dryRun;
+
+  List<BackupFileInfo> logs;
+  List<BackupFileInfo> snaps;
+  List<BackupFileInfo> filesToCopy;
+
+  int mostRecentLogNeededIndex;
+  int snapNeededIndex;
+  int oldestLogNeededIndex;
+
+  /**
+   * Default constructor; requires using parseArgs to setup state.
+   */
+  public RestoreFromBackupTool() {
+    this(null, null, -1L, false);
+  }
+
+  /**
+   * Constructor
+   * @param storageProvider the backup provider from which to restore the backups
+   * @param snapLog the snap and log provider to use
+   * @param zxidToRestore the zxid upto which to restore, or Long.MAX to restore to the latest
+   *                      available transactionally consistent zxid.
+   * @param dryRun whether this is a dryrun in which case no files are actually copied
+   */
+  RestoreFromBackupTool(BackupStorageProvider storageProvider, FileTxnSnapLog snapLog,
+      long zxidToRestore, boolean dryRun) {
+
+    filesToCopy = new ArrayList<BackupFileInfo>();
+    snapNeededIndex = -1;
+    mostRecentLogNeededIndex = -1;
+    oldestLogNeededIndex = -1;
+
+    this.storage = storageProvider;
+    this.zxidToRestore = zxidToRestore;
+    this.snapLog = snapLog;
+    this.dryRun = dryRun;
+  }
+
+  /**
+   * Attempts to perform a restore.
+   */
+  public void run() throws IOException {
+    if (!findFilesToRestore()) {
+      throw new RestoreException("Failed to find a valid snapshot and logs to restore.");
+    }
+
+    // Create a temporary dir in backup storage for backup file processing
+    File tempDirForRestoration =
+        new File(filesToCopy.listIterator().next().getBackedUpFile().getParent(),
```

Comment:
Should we create a local variable called `restorationTmpDir`? and compute it during `findFileseToRestore()`?